src/core/meshcop/joiner.cpp

```diff
@@ -583,8 +641,52 @@ void Joiner::HandleTimer(void)
         break;
     }
 
-    Complete(error);
+    Finish(error);
+}
+
+const char *Joiner::JoinerStateToString(otJoinerState aState)
```

Comment:
I am wondering if we should use template to do this kind of job.
```cpp
template <int kState> const char *Joiner::JoinerStateToString(void);
template <> const char *Joiner::JoinerStateToString< OT_JOINER_STATE_IDLE >(void) { return "Idle"; }
```
This could save code if the string is not used. thoughts @abtink ?

Reply:
Not sure if I follow. How do we use this model?
I think the template argument need to be a compile-time constant. We cannot call it with a variable like `JoineStateToString<aState>`?